README.md
```
                                 __
                                / /________  _________ _____ _____ _
                               / / ___/ __ \/ ___/ __ `/ __ `/ __ `/
                              / (__  ) /_/ (__  ) /_/ / /_/ / /_/ /
                             /_/____/ .___/____/\__,_/\__, /\__,_/
                                   /_/               /____/

                          ⚡ designed for convenience and efficiency ⚡
```

A light-weight lsp plugin based on neovim's built-in lsp with a highly performant UI.

[![](https://img.shields.io/badge/Element-0DBD8B?style=for-the-badge&logo=element&logoColor=white)](https://matrix.to/#/#lspsaga-nvim:matrix.org)

1. [Install](#install)
2. [Example Configuration](#example-configuration)
3. [Usage Tutorial](#usage-tutorial)
4. [Customize Appearance](#customize-appearance)
5. [Backers](#backers)
6. [Donate](#donate)
7. [License](#license)

# Install

you can use some plugins management like `lazy.nvim`, `packer.nvim` to install `lspsaga` and lazyload by the plugin management keyword.

- `cmd` lazyload by `lspsaga` command
- `ft`  lazy.nvim and packer both provide lazyload by filetype. then you can load the lspsaga according the filetypes which you write and use lsp.
- `event` lazyload by event like `BufRead,BufReadPost` make sure your lsp plugin also loaded.
- `dependices` for lazy.nvim you can set the `lspsaga` into `nvim-lspconfig` or other lsp plugin `dependices` keyword.
- `after` for packer you can use after keyword.

## Lazy

```lua
require('lazy').setup({
    'glepnir/lspsaga.nvim',
    event = 'BufRead',
    config = function()
        require('lspsaga').setup({})
    end
},opt)
```

## Packer

```lua
use({
    "glepnir/lspsaga.nvim",
    branch = "main",
    config = function()
        require('lspsaga').setup({})
    end,
})
```

# Example Configuration

```lua
require('lazy').setup({
    'glepnir/lspsaga.nvim',
    event = 'BufRead',
    config = function()
      require('lspsaga').setup({})
    end
})

local keymap = vim.keymap.set
-- Lsp finder find the symbol definition implement reference
-- if there is no implement it will hide
-- when you use action in finder like open vsplit then you can
-- use <C-t> to jump back
keymap("n", "gh", "<cmd>Lspsaga lsp_finder<CR>")

-- Code action
keymap({"n","v"}, "<leader>ca", "<cmd>Lspsaga code_action<CR>")

-- Rename
keymap("n", "gr", "<cmd>Lspsaga rename<CR>")

-- Peek Definition
-- you can edit the definition file in this float window
-- also support open/vsplit/etc operation check definition_action_keys
-- support tagstack C-t jump back
keymap("n", "gd", "<cmd>Lspsaga peek_definition<CR>")

-- Go to Definition
keymap("n","gd", "<cmd>Lspsaga goto_definition<CR>")

-- Show line diagnostics you can pass argument ++unfocus to make
-- show_line_diagnostics float window unfocus
keymap("n", "<leader>sl", "<cmd>Lspsaga show_line_diagnostics<CR>")

-- Show cursor diagnostic
-- also like show_line_diagnostics  support pass ++unfocus
keymap("n", "<leader>sc", "<cmd>Lspsaga show_cursor_diagnostics<CR>")

-- Show buffer diagnostic
keymap("n", "<leader>sb", "<cmd>Lspsaga show_buf_diagnostics<CR>")

-- Diagnostic jump can use `<c-o>` to jump back
keymap("n", "[e", "<cmd>Lspsaga diagnostic_jump_prev<CR>")
keymap("n", "]e", "<cmd>Lspsaga diagnostic_jump_next<CR>")

-- Diagnostic jump with filter like Only jump to error
keymap("n", "[E", function()
  require("lspsaga.diagnostic").goto_prev({ severity = vim.diagnostic.severity.ERROR })
end)
keymap("n", "]E", function()
  require("lspsaga.diagnostic").goto_next({ severity = vim.diagnostic.severity.ERROR })
end)

-- Toggle Outline
keymap("n","<leader>o", "<cmd>Lspsaga outline<CR>")

-- Hover Doc
-- if there has no hover will have a notify no information available
-- to disable it just Lspsaga hover_doc ++quiet
-- press twice it will jump into hover window
keymap("n", "K", "<cmd>Lspsaga hover_doc<CR>")
-- if you want keep hover window in right top you can use ++keep arg
-- notice if you use hover with ++keep you press this keymap it will
-- close the hover window .if you want jump to hover window must use
-- wincmd command <C-w>w
keymap("n", "K", "<cmd>Lspsaga hover_doc ++keep<CR>")

-- Callhierarchy
keymap("n", "<Leader>ci", "<cmd>Lspsaga incoming_calls<CR>")
keymap("n", "<Leader>co", "<cmd>Lspsaga outgoing_calls<CR>")

-- Float terminal
keymap({"n", "t"}, "<A-d>", "<cmd>Lspsaga term_toggle<CR>")
```

## Usage Tutorial

**Notice that title in float window must need neovim version >= 0.9**
**So if you are using the nvim 0.8 version you can't see title**
**If you are using nvim 0.9 and want disable title see [#customize-appearance]

**For some new you don't need copy all options into setup function. just set the options that you**
**changed into setup function it will extend to default options!**

you can find these doc in neovim by `:h lspsaga` . 

## default options

```lua
  preview = {
    lines_above = 0,
    lines_below = 10,
  },
  scroll_preview = {
    scroll_down = '<C-f>',
    scroll_up = '<C-b>',
  },
  request_timeout = 2000,
```


## :Lspsaga lsp_finder

`Finder` to show the defintion,reference,implement(only show when current word is interface or some type)

default finder options

```lua
  finder = {
    edit = { 'o', '<CR>' },
    vsplit = 's',
    split = 'i',
    tabe = 't',
    quit = { 'q', '<ESC>' },
  },
```

<details>
<summary>lsp finder show case</summary>

<img
src="https://user-images.githubusercontent.com/41671631/212032702-f45bba5a-3e2e-465d-85c3-3d02d1b88da4.gif" height=80% width=80%/>
</details>


## :Lspsaga peek_definition

there has two commands `Lspsaga peek_definition` and `Lspsaga goto_definition`, the `peek_definition` work like vscode that show the target file in a floatwindow you can edit as normalize.

options with default value

```lua
  definition = {
    edit = '<C-c>o',
    vsplit = '<C-c>v',
    split = '<C-c>i',
    tabe = '<C-c>t',
    quit = 'q',
    close = '<Esc>',
  }
```

<details>
<summary>peek_definition show case</summary>

the step in this gif show case
- `gd` run `lspsaga peek_definition`.
- do some comment edit then `:w` to save.
- `<C-c> o` jump to this file.
- lspsaga will show becacon highlight after jump .

<img
src="https://user-images.githubusercontent.com/41671631/212002926-60c11060-233c-4610-a86e-57decefe6927.gif" height=80% width=80%/>
</details>

## :Lspsaga goto_definition

jump to definition and show beacon


## :Lspsaga code_action

options with default value

```lua
  code_action = {
    num_shortcut = true,
    keys = {
      -- string |table type
      quit = 'q',
      exec = '<CR>',
    },
  },
```
- `num_shortcut` it's `true` by default then you can use number to fast run action

<details>
<summary> code_action show case </summary>

- `ga` run `Lspsaga code_action`
- `j` to move and show the code action preview
- `<Cr>` to run a action
<img src="https://user-images.githubusercontent.com/41671631/212005522-bc7fa99b-6c6f-4c0e-b7fc-c95edee5c169.gif" height=80% width=80%/>
</details>

## :Lspsaga Lightbulb

when there has code action it will show a lightbulb, default options.

```lua
  lightbulb = {
    enable = true,
    enable_in_insert = true,
    sign = true,
    sign_priority = 40,
    virtual_text = true,
  },
```

<details>
<summary> lightbulb show case </summary>
<img src="https://user-images.githubusercontent.com/41671631/212009221-e0fb193e-f69d-4ed6-a4a2-d9ecb589f211.gif" height=80% width=80%/>
</details>

## :Lspasga hover_doc

lspsaga use treesitter markdown parser to render hover. so you must install markdown parser.
you can press shotcut of `Lspsaga hover_doc` twice jump into the hover window. in my case is `K`.

<details>
<summary> hover_doc show case </summary>

- `K` to run `Lspsaga hover_doc`.
- press `K` again jump into the hover window.
- `q` to quit.

<img src="https://user-images.githubusercontent.com/41671631/212010765-55341ba1-95c2-41e9-b4bd-03827676ee94.gif" height=80% width=80%/>

</details>

## :Lspsaga diagnostic_jump_next

jump to next diagnostic position then show beacon and show the codeaction. default options

```lua
  diagnostic = {
    show_code_action = true,
    show_source = true,
    jump_num_shortcut = true,
    keys = {
      exec_action = 'o',
      quit = 'q',
      go_action = 'g'
    },
  },
```

- `go_action` can quickly jump to action line in diagnostic floatwindow
- `jump_num_shortcut` default is true after jump. Lspasga will auto bind the numbers to action index
then just press number will execute the code action. after window close these number map will delete

also you can use a filter in diagnostic jump by using lspsaga function. function params is a table
same as `:h vim.diagnostic.get_next`

```lua
-- this is mean only jump to error position
-- or goto_next
require("lspsaga.diagnostic").goto_prev({ severity = vim.diagnostic.severity.ERROR })
```

<details>
<summary>diagnostic jump show case </summary>

- `[e` to jump next diagnostic position and show beacon and code actions in diagnostic window
- press number `2` to execute code action no need jump into that window

<img src="https://user-images.githubusercontent.com/41671631/212669236-dda9f06b-6840-41cd-92b4-5c2290077c37.gif" height=80% width=80%/>

- if you want to see the code action. you can using `<C-w>w` jump into the floatwindow
- `g` go to the action line see the code action preview
- `o` execute the action

<img src="https://user-images.githubusercontent.com/41671631/212669236-dda9f06b-6840-41cd-92b4-5c2290077c37.gif" height=80% width=80%/>

</details>

## :Lspsaga show_diagnostic

`show_line_diagnostics` , `show_buf_diagnostics` ,`show_cursor_diagnostics`

<details>
<summary> show diagnostics show case </summary>
<img src="https://user-images.githubusercontent.com/41671631/212220793-a52215fd-5f60-4be6-8132-78247b921f1e.gif" height=80% width=80%/>
</details>

## :Lspsaga rename

lsp rename with select. default options

```lua

  rename = {
    quit = '<C-c>',
    exec = '<CR>',
    mark = 'x',
    confirm = '<CR>',
    in_select = true,
    whole_project = true,
  },
```

- `whole_project` support rename whole project after using lsp rename by `rg`.
- `mark` after lsp rename if there still have this word in project then will popup a window that include the files which include the word.then you can use this key mark the line that you want changed.
- `confirm` when you marked the lines then press this keymap to execute.

<details>
<summary> rename show case</summary>

- `gr` to run `Lspsaga rename`
- `stesdd<CR>` 

<img src="https://user-images.githubusercontent.com/41671631/212015791-5a278ace-d23a-4954-bb95-1978f51153a7.gif" height=80% width=80%/>
</details>

## :Lspsaga outline

default options

```lua
  outline = {
    win_position = 'right',
    win_with = '',
    win_width = 30,
    show_detail = true,
    auto_preview = true,
    auto_refresh = true,
    auto_close = true,
    custom_sort = nil,
    keys = {
      jump = 'o',
      expand_collapse = 'u',
      quit = 'q',
    },
  },
```

<details>
<summary>outline show case</summary>

- `<Leader>o` run `Lspsaga outline`
- `j` move down
- `o` to jump

<img src="https://user-images.githubusercontent.com/41671631/212017018-6753e470-58e4-498e-8812-5ff416ff27c1.gif" height=80% width=80%/>
</details>

## :Lspsaga incoming_calls

run lsp callhierarchy incoming_calls.  default options

```lua
  callhierarchy = {
    show_detail = false,
    keys = {
      edit = 'e',
      vsplit = 's',
      split = 'i',
      tabe = 't',
      jump = 'o',
      quit = 'q',
      expand_collapse = 'u',
    },
  },
```

<details>
<summary>incoming_calls show case</summary>
<img src="https://user-images.githubusercontent.com/41671631/212018219-26ed4a5f-00e1-488a-8a87-1a89f2c5d14b.gif" height=80% width=80%/>
</details>

## :Lspsaga outgoing_calls

run lsp callhierarchy outgoing_calls

<details>
<summary>outgoing show_case</summary>
<img src="https://user-images.githubusercontent.com/41671631/212024418-cf26f3f7-7acb-46df-a50a-9abe3f8f68f3.gif" height=80% width=80%/>
</details>

## :Lspsaga symbols in winbar

require your neovim version >= 0.8. options with default value

```lua
  symbol_in_winbar = {
    enable = true,
    separator = ' ',
    hide_keyword = true,
    show_file = true,
    folder_level = 2,
    respect_root = false,
    color_mode = true,
  },
```
- `hide_keyword` default is true it will hide some keyword or tmp variable make symbols more clean
- `folder_level` work with `show_file`
- `respect_root` will respect lsp root if this is true will ignore the folder_level. if there is no
  lsp client usage then fallback to folder_level
- `color_mode` default is true, false mean only icon have color.

<details>
<summary>symbols in winbar </summary>
<img src="https://user-images.githubusercontent.com/41671631/212026278-11012b17-209c-4b55-b76c-1c3d8d9a2eb2.gif" height=80% width=80%/>
</details>

## :Lspsaga symbols in custom winbar/statusline

lspsaga provide an api that you can use in your custom winbar or statusline.

```lua
vim.wo.winbar/ vim.wo.stl = require('lspsaga.symbolwinbar'):get_winbar()
```

## :Lspsaga term_toggle

simple floaterm

<details>
<summary> float terminal toggle</summary>
<img src="https://user-images.githubusercontent.com/41671631/212027060-56d1cebc-c6a8-412e-bd01-620aac3029ed.gif" height=80% width=80%/>
</details>

## Customize Appearance

## :Lspsaga UI

default ui options

```lua
  ui = {
    -- currently only round theme
    theme = 'round',
    -- this option only work in neovim 0.9
    title = true,
    -- border type can be single,double,rounded,solid,shadow.
    border = 'solid',
    winblend = 0,
    expand = '',
    collapse = '',
    preview = ' ',
    code_action = '💡',
    diagnostic = '🐞',
    incoming = ' ',
    outgoing = ' ',
    colors = {
      --float window normal background color
      normal_bg = '#1d1536',
      --title background color
      title_bg = '#afd700',
      red = '#e95678',
      magenta = '#b33076',
      orange = '#FF8700',
      yellow = '#f7bb3b',
      green = '#afd700',
      cyan = '#36d0e0',
      blue = '#61afef',
      purple = '#CBA6F7',
      white = '#d1d4cf',
      black = '#1c1c19',
    },
    kind = {},
  },
```

# Custom Highlight

If you don't like config the colors or you use switch colorschem according time then you can use 
the old way. change the highlight group in your colorscheme or config.
find all highlight groups in [highlight.lua](./lua/lspsaga/highlight.lua)

# Custom Kind

through `ui.kind` to modified the all kinds defined in [lspkind.lua](./lua/lspsaga/lspkind.lua)
the key in `ui.kind` is kind name, value can be string or table if string type is icon if table type
is `{icon, color}`.

# Backers
Thanks for all.

[@Möller Lukas](https://github.com/lmllrjr)
[@HendrikPetertje](https://github.com/HendrikPetertje)
[@Bojan Wilytsch](https://github.com/bwilytsch)
[@Burgess Darrion](https://github.com/ca-mantis-shrimp)

# Donate

[![](https://img.shields.io/badge/PayPal-00457C?style=for-the-badge&logo=paypal&logoColor=white)](https://paypal.me/bobbyhub)

Currently I need some donate. If you'd like to support my work financially, donate through [paypal](https://paypal.me/bobbyhub).
Thanks!

# License

Licensed under the [MIT](./LICENSE) license.


Changelog.md
## Version 0.2.3 -- 2023-01-13

- Refactor all

## Version 0.2.2 -- 2022--09-09

- fix symbolwinbar high memory usage
- `preview_definition` chang to `peek_definition` support edit definition file in floatwindow and highlight
- `code_action` support in visual mode .
- `finder` preview highlight current word.
-  show diagnostic source by default.
- add lspsaga codeaction keymap in diagnostic header

## Remove

-  remove `definition_preview_icon`  option
- remove `finder_preview_hl_ns`
- remove `diagnostic_source_bracket`
- remove `show_diagnostic_source` . 

## Version 0.2.1 -- 2022-08-30

### New

- improve scroll in preview remove old implement of preview
- add new option  `scroll_in_preview` use default scroll keymap to scroll `hover` `finder preview`
- use `CursorMoved` with timer in LightBulb instead of using `CursorHold` `CursorHoldI`
- new option `update_time` in code_action_lightbulb
- auto jump into the `preview_definition` window
- `floaterm` can save the state.

### Bug fix

- fix `definition_preview` and `hover` window not closed when `CursorMoved`
- fix `finder` works not well when server has multiple servers

### Remove

- remove function `action.smart_scroll_with_saga(1)` no need this function

## Version 0.2

### BreakChange

Please use command in `vim.keymap.set` rhs . check the README.

### New

- `Finder:` support show `implement` in `Lspsaga lsp_finder`
- `Finder:` add option `imp` in `finder_icons`
- `LightBulb` add api to genreate sign need neovim 0.8 nightly. 0.7 version also use `vim.fn.sign_place/unplace`
-  `LightBulb` set the `CursorHold CursorHoldI` only take effect when the buffer has lsp server. and remove them when 
    buffer delete ,keep the autocmds clean
- `Background of Lspsaga floatwindow` now it will use the colorscheme Normal highlight
- `Notify` add notify for some commands if there has no server give a message
-  `Symbolwinbar` when delete the buffer remove the buffer events that symbolwinbar used,keep autocmds clean
-  `Outline` remove the patch as [neovim#19458](https://github.com/neovim/neovim/issues/19458#)completed
-  `Outline` fix bug
-  `Definition` rewrite definition
-  Close the `finder` `rename` window when use wincmd shortcut jump to other window like `<C-w>h`
-  `Diagnostic` fix the virtual text bug when jump diagnostic
-   Better `show_line_diagnostic` and `show_cursor_diagnostic`
-  `custom_kind` option change the default lspkind icon and color

### Remove

- remove `Lspsaga implement`.


lua/lspsaga/init.lua
local api = vim.api
local saga = {}
saga.saga_augroup = api.nvim_create_augroup('Lspsaga', { clear = true })

local default_config = {
  ui = {
    theme = 'round',
    border = 'solid',
    title = true,
    winblend = 0,
    expand = '',
    collapse = '',
    preview = ' ',
    code_action = '💡',
    diagnostic = '🐞',
    incoming = ' ',
    outgoing = ' ',
    colors = {
      --float window normal bakcground color
      normal_bg = '#1d1536',
      --title background color
      title_bg = '#afd700',
    },
    kind = {},
  },
  diagnostic = {
    show_code_action = true,
    show_source = true,
    jump_num_shortcut = true,
    keys = {
      exec_action = 'o',
      quit = 'q',
      go_action = 'g',
    },
  },
  code_action = {
    num_shortcut = true,
    keys = {
      quit = 'q',
      exec = '<CR>',
    },
  },
  lightbulb = {
    enable = true,
    enable_in_insert = true,
    -- cache_code_action = true,
    sign = true,
    sign_priority = 40,
    virtual_text = true,
  },
  preview = {
    lines_above = 0,
    lines_below = 10,
  },
  scroll_preview = {
    scroll_down = '<C-f>',
    scroll_up = '<C-b>',
  },
  request_timeout = 2000,
  finder = {
    edit = { 'o', '<CR>' },
    vsplit = 's',
    split = 'i',
    tabe = 't',
    quit = { 'q', '<ESC>' },
  },
  definition = {
    edit = '<C-c>o',
    vsplit = '<C-c>v',
    split = '<C-c>i',
    tabe = '<C-c>t',
    quit = 'q',
    close = '<Esc>',
  },
  rename = {
    quit = '<C-c>',
    exec = '<CR>',
    mark = 'x',
    confirm = '<CR>',
    in_select = true,
    whole_project = true,
  },
  symbol_in_winbar = {
    enable = true,
    separator = ' ',
    hide_keyword = true,
    show_file = true,
    folder_level = 2,
    respect_root = false,
    color_mode = true,
  },
  outline = {
    win_position = 'right',
    win_with = '',
    win_width = 30,
    show_detail = true,
    auto_preview = true,
    auto_refresh = true,
    auto_close = true,
    custom_sort = nil,
    keys = {
      jump = 'o',
      expand_collapse = 'u',
      quit = 'q',
    },
  },
  callhierarchy = {
    show_detail = false,
    keys = {
      edit = 'e',
      vsplit = 's',
      split = 'i',
      tabe = 't',
      jump = 'o',
      quit = 'q',
      expand_collapse = 'u',
    },
  },
  server_filetype_map = {},
}

function saga.theme()
  local theme = {
    ['round'] = {
      left = '',
      right = '',
    },
  }

  return theme[saga.config.ui.theme]
end

function saga.setup(opts)
  opts = opts or {}
  saga.config = vim.tbl_deep_extend('force', default_config, opts)

  require('lspsaga.highlight'):init_highlight()
  if saga.config.lightbulb.enable then
    require('lspsaga.lightbulb').lb_autocmd()
  end

  if saga.config.symbol_in_winbar.enable or saga.config.symbol_in_winbar.in_custom then
    require('lspsaga.symbolwinbar'):symbol_autocmd()
  end
end

---@deprecated
function saga.init_lsp_saga()
  vim.notify(
    'lspsaga.nvim v0.2.3+ has breaking changes. Please read the docs and migrate your configuration to the new "setup" function!',
    vim.log.levels.ERROR
  )
end

return saga
